{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A summary of cattle (*Bos taurus*) assemblies\n",
    "There are 3 assemblies available for cattle, Btau, UMD and ARS-UCD. ARS-UCD is now available on [NCBI](ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Bos_taurus/latest_assembly_versions/GCF_002263795.1_ARS-UCD1.2). NCBI also offer [downloads for older assemblies of Btau and UMD](ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Bos_taurus/all_assembly_versions/suppressed/).\n",
    "\n",
    "Above links were tracked from [this very brief history of cattle assemblies](https://bovinegenome.elsiklab.missouri.edu/history).\n",
    "\n",
    "### Btau\n",
    "This was assembled by BCM-HGSC (Baylor College of Medicine Human Genome Sequecing Center)\n",
    "\n",
    "| Date Released | Release Name | Coverage | Comments |\n",
    "| -- | -- | -- | -- |\n",
    "| December 2015 | Btau_5.0.1 | 25x | Improved draft assembly combined 19x PacBio data with UMD 3.1 using PBJelly. |\n",
    "| July 2012 | Btau_4.6.1 | 7.1x | Draft assembly replaced data in Btau_4.5 with high quality finished data where available. |\n",
    "| October 2009 | Btau_4.5 | 7.1x | Draft assembly incorporating additional Whole Genome Shotgun (WGS) contigs from Btau_2.0 into Btau_4.0 |\n",
    "| April 2009 | Btau_4.2 | 7.1x | Draft assembly Btau_4.0 data replaced with high quality finished sequence where available. |\n",
    "| October 2007 | Btau_4.0 | 7.1x | Draft assembly using (WGS) reads from small insert clones and BAC sequences. Mapped to chromosomes using refin\\\n",
    "ed mapping information. |\n",
    "| August 2006 | Btau_3.1 | 7.1x | Draft assembly (WGS) reads from small insert clones and BAC sequences. |\n",
    "| March 2005 | Btau_2.0 | 6.2x | Preliminary assembly using (WGS) reads from small insert clones and BAC end sequences. |\n",
    "| September 2004 | Btau_1.0 | 3x | Preliminary assembly using (WGS) reads from small insert clones. |\n",
    "\n",
    "- Refer: https://www.hgsc.bcm.edu/other-mammals/bovine-genome-project\n",
    "\n",
    "### UMD\n",
    "| Date Released | Release Name |\n",
    "| -- | -- |\n",
    "| 2014-11-25 | UMD 3.1.1 |\n",
    "| 2009-12-1 | UMD 3.1 |\n",
    "\n",
    "### ARS-UCD\n",
    "[Available on NCBI](ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Bos_taurus/latest_assembly_versions/GCF_002263795.1_ARS-UCD1.2). This was released on 2018-04-11. [This text file](ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Bos_taurus/latest_assembly_versions/GCF_002263795.1_ARS-UCD1.2/GCF_002263795.1_ARS-UCD1.2_assembly_stats.txt) states the assembly statistics.\n",
    "\n",
    "### Other notes\n",
    "- 2009 major releases\n",
    "  - UMD 2.0, by The Center for Bioinformatics Computational Biology at the University of Maryland. [Source abstract](https://jcea.agr.hr/articles/773128_The_new_bovine_reference_genome_assembly_provides_new_insight_into_genomic_organization_of_the_bovine_major_histocompatibi_en.pdf).\n",
    "  - Btau 4.0, by BCM-HGSC (Baylor College of Medicine Human Genome Sequecing Center)\n",
    "- 2011\n",
    "  - NCBI designated UMD3.1 as the reference assembly.\n",
    "  - Btau was set as alternative"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do we need to map the Illumina manifests to ARS-UCD 1.2?\n",
    "- Illumina BovineHD has both UMD3.1 and Btau4.2 maps.\n",
    "- My previous experience shows that this will increase imputation results.\n",
    "- We have so many chip platforms, imputation is not avoidable."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Checklist of files for new manifest\n",
    "- [x] [Refererence ARS-UCD 1.2](ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Bos_taurus/latest_assembly_versions/GCF_002263795.1_ARS-UCD1.2/GCF_002263795.1_ARS-UCD1.2_genomic.fna.gz).\n",
    "  - It is also possible to download [ARS-ucd 1.2, Btau 5.0.1, and Bos taurus UMD 3.1.1 all together](https://www.ncbi.nlm.nih.gov/assembly/?term=Bos+taurus), which seems not necessary.\n",
    "- [x] SNP manifests\n",
    "  - [x] Illumina BovineSNP50 v3.0 has 2 manifest files, namely A1, and A2.\n",
    "    - Obviously A2 is better than A1. A2 has 183 SNP of chromosome 0, while A1 has 622 unassigned SNP.\n",
    "    - A2 was created on May 10, 2016. \n",
    "    - A1 was created on Jan 27, 2016.\n",
    "    - I will use [A2](https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/bovinesnp50/v3/bovinesnp50-v3-a2-manifest-file-csv.zip)\n",
    "  - [x] [BovineHD](ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/productfiles/bovinehd/bovinehd-b1-annotation-file.zip). Note there is also two other manifest files created in 2014. I use the suport file created on Feb 23, 2016. This was mapped against UMD3.1."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A comparison of various platforms\n",
    "- Note:\n",
    "  - **N** = Norway\n",
    "  - **D** = Dutch\n",
    "  - **G** = Germany\n",
    "  - **I** = Illumina\n",
    "\n",
    "### Overview\n",
    "- The Norwegian dataset is only on one HD SNP platform.\n",
    "- The Dutch dataset are on 10 different platforms.\n",
    "  - 9066, 10690, 10993, 11483, 50905\n",
    "  - 52384, 53218, 54001, 54609, 777962\n",
    "- The German data are on 2 platforms. The number of SNP shows that they are a subset of the Dutch 10 platforms.\n",
    "  - 53218, 54609\n",
    "  - **But** it has no map information at the moment.\n",
    "    - [x] Actually are Illumina 50k v2 and v3\n",
    "    - New data received on April 4, 2020\n",
    "- Verification by SNP names\n",
    "  - **D**-777692 $\\equiv$ **I**-HD\n",
    "  - **D**-53218 $\\equiv$ **I**-50.\n",
    "- From Illumina, searching *bovine* and *manifest* only gives HD=777692, and 50v3=53218 two platforms.\n",
    "\n",
    "### Comparisons\n",
    "Note:\n",
    "- http://bioinfo.genotoul.fr/jvenn/ may be a good tool to plot venn diagrams.\n",
    "- But we have too many classes/lists here.\n",
    "- Also:\n",
    "  - **D**-777692 $\\equiv$ **I**-HD\n",
    "  - **D**-53218 $\\equiv$ **I**-50.\n",
    "\n",
    "#### The two Illumina chips\n",
    "- HD and 50k chips share 48,492 SNP\n",
    "- [ ] maybe remvoe sex and mitochondrian chromosome first?\n",
    "\n",
    "#### German data\n",
    "If the two platforms are the same as the Dutch 54609, and 53218, then\n",
    "- they share 50,908 SNP. 50k+ SNP are good for later anaylysis.\n",
    "- together with **I**-HD, 46,290 SNP shared\n",
    "\n",
    "#### Norwegian data\n",
    "- Share 620,758 with **I**-HD\n",
    "- Share 41,437 with **I**-50k\n",
    "- Share with **I**-HD, and two German chips: 39,582\n",
    "  - This is still acceptable.\n",
    "  - Will be used for the Dutch data comparison.\n",
    "- Many loci may have been filtered out (???)\n",
    "- Its 1,421 special loci mean it is not a Illumina HD (???)\n",
    "\n",
    "#### The Dutch data\n",
    "- is a headache\n",
    "- I use 39,582 to check the rest 8 chips\n",
    "- below were only on the SNP names\n",
    "  - a *manifest* file, i.e. SNP definition may help\n",
    "  - a same SNP may have different names.\n",
    "\n",
    "| Platform | $N_{\\textrm{shared SNP}}$ | Score |\n",
    "| --: | --: | :--: |\n",
    "| 54,001 | 38,528 | :smiley: |\n",
    "| 52,384 | 8,230 | :disappointed: |\n",
    "| 50,905 | 15,688 | :disappointed: |\n",
    "| 11,483 | 6,412 | :frowning: |\n",
    "| 10,993 | 6,405 | :frowning: |\n",
    "| 10,690 | 6,341 | :frowning: |\n",
    "| 9,066 | 6,373 | :smiley: |\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data management\n",
    "### To date 22/01/2020 11:16\n",
    "#### German\n",
    "\n",
    "| Platform | $N_{\\mathrm{ID}}$ |\n",
    "| --: | --: |\n",
    "| 53,218 | 739 |\n",
    "| 54,609 | 77 |\n",
    "| Total | 816 |\n",
    "\n",
    "#### Norge\n",
    "\n",
    "| Chip type | $N_{\\mathrm{ID}}$ |\n",
    "| --: | --: |\n",
    "| 622,179 | 63,721 |\n",
    "\n",
    "#### Dutch\n",
    "\n",
    "| Platform | $N_{\\mathrm{ID}}$ |\n",
    "| --: | --: |\n",
    "| 9,066 | 4 |\n",
    "| 10,690 | 1,994 |\n",
    "| 10,993 | 103 |\n",
    "| 11,483 | 109 |\n",
    "| 50,905 | 10 |\n",
    "| 52,384 | 6 |\n",
    "| 53,218 | 44 |\n",
    "| 54,001 | 1 |\n",
    "| 54,609 | 343 |\n",
    "| 777,962 | 10 |\n",
    "| Total | 2,623 |\n",
    "\n",
    "- Only 10690, 10993, 11483, v2, v3, and HD `6` platforms were used for downstream analysis\n",
    "\n",
    "### More data needed\n",
    "- Norway\n",
    "  - [x] More genotype data? **No**\n",
    "  - [x] SNP platform description **No**\n",
    "  - [x] Raw genotype file **Done**\n",
    "    - not filtered, not imputed\n",
    "    - as map may be upgraded\n",
    "- German\n",
    "  - [x] More genotypes. **No**\n",
    "  - [x] SNP information. **No**\n",
    "    - no before\n",
    "- The Netherland\n",
    "  - [x] More genotypes. **No**\n",
    "  - [x] SNP platform manifests. **Illumina**\n",
    "    - 52384, 50905, are so disappointing for comparison between platforms.\n",
    "    - 11483, 10993, 10690 are not good either based on SNP names\n",
    "\n",
    "### Genotype storage\n",
    "- Plink genotype management seems very fast, storage efficient and convenient.\n",
    "- Different format can be usually converted anyway.\n",
    "\n",
    "### Map updates\n",
    "- [x] Map SNP onto ARS-UCD. **Later**\n",
    "- [x] Update new map with plink? **Later**\n",
    "- It is very fast to deal with anyway."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.0-DEV",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
