{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Roadmap\n",
    "- [x] Totally convert the project into a Julia package framework\n",
    "- [x] Filter all the genotypes\n",
    "  - Discard chromosome 0, X, Y, MT, and SNP not on platform 50k-v3\n",
    "- [x] Update map to Illumina 50k v3\n",
    "- [ ] QC about this subset\n",
    "- [ ] Test imputation quality\n",
    "- [ ] Impute and do genomic prediction with GBLUP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The ReDiverse.jl package\n",
    "\n",
    "I totally changed my jobs for this project into the life cycle of a Julia package development. This made, I feel, the work more fluent and effective.\n",
    "\n",
    "To obtain a local copy of my code, one can:\n",
    "```bash\n",
    "cd a-dir-you-want-to-hold-these codes\n",
    "git clone https://github.com/xijiang/ReDiverse.jl ReDiverse\n",
    "cd ReDiverse\n",
    "julia # to enter julia REPL if you want run the codes\n",
    "```\n",
    "```julia\n",
    "]  # to enter julia package management\n",
    "add Plots, GR, LaTeXStrings, Test, Printf  # if not installed\n",
    "activate .\n",
    "test  # this will give you a lot error messages, as the data were private\n",
    "```\n",
    "\n",
    "You can:\n",
    "```bash\n",
    "grep function src/*jl\n",
    "```\n",
    "to see what functions I wrote for this project. Then in REPL:\n",
    "```julia\n",
    "?function-name\n",
    "```\n",
    "to see issues and considerations, or documentation, for the function. I have written a lot documentations for almost every function.\n",
    "\n",
    "Later, when the package repository is public again, you just go back to this folder and:\n",
    "```bash\n",
    "git pull\n",
    "```\n",
    "to update new codes and results.\n",
    "You might also want to:\n",
    "```bash\n",
    "jupyter-labextension install @jupyterlab/toc\n",
    "```\n",
    "To have a table of content side bar to view my reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract genotypes for downstream anaysis\n",
    "### Check map concordance\n",
    "Map 50k v3 has 51730 autosomal SNP\n",
    "Below will count each map that are not 50kv3. I will refer this as **Table 1** later.\n",
    "1. Total number of SNP\n",
    "2. number of SNP that are also in 50kv3\n",
    "3. of 2, number of autosomal SNP\n",
    "4. of 2, number of non-autosomal SNP\n",
    "5. of 2, has same chromosome number and bp position\n",
    "\n",
    "**Table1**\n",
    "\n",
    "|   Map |     1. |    2. |    3. |   4. |    5. |\n",
    "| --: | --: | --: | --: | --: | --: |\n",
    "| 50kv1 |  58276 | 47877 | 46727 | 1150 |     0 |\n",
    "| 50kv2 |  58763 | 49465 | 49069 |  396 | 49048 |\n",
    "|  777k | 777962 | 47155 | 46796 |  359 | 46777 |\n",
    "| 10690 |  10690 |  7465 |  7461 |    4 |  7419 |\n",
    "| 10993 |  10993 |  7539 |  7535 |    4 |  7493 |\n",
    "| 11483 |  11483 |  7549 |  7545 |    4 |  7503 |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obviously, map 50k-v1 has changed quite a lot. It has 1150 loci that were mapped to non-autosomes that are now on autosomes. And the positions are totally changed. \n",
    "The conclusion is I need to **update the maps**\n",
    "There are also ~3-400 SNP discrepancy between 50k-v2, 777k and 50k-v3. So the following question is whether to use 50k-v3 or 777k map.\n",
    "Since [BovineHD](ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/productfiles/bovinehd/bovinehd-b1-annotation-file.zip) was created in 2014, [50k v3](https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/bovinesnp50/v3/bovinesnp50-v3-a2-manifest-file-csv.zip) was created in 2016, I will use 50k-v3 as the target map.\n",
    "\n",
    "- [ ] I am also considering to map these SNP with the most recent genome reference\n",
    "  - If I have time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update the 5 maps of the 6 platforms\n",
    "- [x] Extract the SNPs in column **3.** of **Table 1**\n",
    "  - Thre rest SNP in these 5 maps will discarded.\n",
    "  - The Norwegian data has no result from 50k-v3 platform. They can't be imputed to 51730 level of 50k-v3, as data from other countries can't be used as references. Let 777k SNP set as $v_7$, the Norwegian SNP set $s = \\cup(v_1, v_2, v_7)$, then the SNP in $\\cap(s, v_3)$ will be the final target number of SNP to impute to.\n",
    "  - Thus the shared number of loci will be even smaller. I summerized the final number of shared loci in **Table 2**\n",
    "- [ ] Update the map with 50k-v3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Table 2**\n",
    "\n",
    "| Target | 50k-V1 | 50k-V2 |    HD | PF-10690 | PF-10993 | PF-11483 |\n",
    "| --: | --: | --: | --: | --: | --: | --: |\n",
    "|  51626 |  47877 |  49465 | 47155 |     7364 |     7438 |     7448 |\n",
    "\n",
    "Platform 50k v3 has 51730 autosomal SNP. Here I found 51626 usable. There is not a big loss at this stage."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new plink file set\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.0-DEV",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
