{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Roadmap\n",
    "- [x] Totally convert the project into a Julia package framework\n",
    "- [x] Filter all the genotypes\n",
    "  - Discard chromosome 0, X, Y, MT, and SNP not on platform 50k-v3\n",
    "- [x] Update map to Illumina 50k v3\n",
    "- ~~QC about this subset~~ This result should be **ignored**\n",
    "- [x] Merge samples by country\n",
    "  - only 3 sets finally"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The ReDiverse.jl package\n",
    "\n",
    "I totally changed my jobs for this project into the life cycle of a Julia package development. This made, I feel, the work more fluent and effective.\n",
    "\n",
    "To obtain a local copy of my code, one can:\n",
    "```bash\n",
    "cd a-dir-you-want-to-hold-these codes\n",
    "git clone https://github.com/xijiang/ReDiverse.jl ReDiverse\n",
    "cd ReDiverse\n",
    "julia # to enter julia REPL if you want run the codes\n",
    "```\n",
    "```julia\n",
    "]  # to enter julia package management\n",
    "add Plots, GR, LaTeXStrings, Test, Printf  # if not installed\n",
    "activate .\n",
    "test  # this will give you a lot error messages, as the data were private\n",
    "```\n",
    "\n",
    "You can:\n",
    "```bash\n",
    "grep function src/*jl\n",
    "```\n",
    "to see what functions I wrote for this project. Then in REPL:\n",
    "```julia\n",
    "?function-name\n",
    "```\n",
    "to see issues and considerations, or documentation, for the function. I have written a lot documentations for almost every function.\n",
    "\n",
    "Later, when the package repository is public again, you just go back to this folder and:\n",
    "```bash\n",
    "git pull\n",
    "```\n",
    "to update new codes and results.\n",
    "You might also want to:\n",
    "```bash\n",
    "jupyter-labextension install @jupyterlab/toc\n",
    "```\n",
    "To have a table of content side bar to view my reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract genotypes for downstream anaysis\n",
    "### Check map concordance\n",
    "Map 50k v3 has 51730 autosomal SNP\n",
    "Below will count each map that are not 50kv3. I will refer this as **Table 1** later.\n",
    "1. Total number of SNP\n",
    "2. number of SNP that are also in 50kv3\n",
    "3. of 2, number of autosomal SNP\n",
    "4. of 2, number of non-autosomal SNP\n",
    "5. of 2, has same chromosome number and bp position\n",
    "\n",
    "**Table1**\n",
    "\n",
    "|   Map |     1. |    2. |    3. |   4. |    5. |\n",
    "| --: | --: | --: | --: | --: | --: |\n",
    "| 50kv1 |  58276 | 47877 | 46727 | 1150 |     0 |\n",
    "| 50kv2 |  58763 | 49465 | 49069 |  396 | 49048 |\n",
    "|  777k | 777962 | 47155 | 46796 |  359 | 46777 |\n",
    "| 10690 |  10690 |  7465 |  7461 |    4 |  7419 |\n",
    "| 10993 |  10993 |  7539 |  7535 |    4 |  7493 |\n",
    "| 11483 |  11483 |  7549 |  7545 |    4 |  7503 |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obviously, map 50k-v1 has changed quite a lot. It has 1150 loci that were mapped to non-autosomes that are now on autosomes. And the positions are totally changed. \n",
    "The conclusion is I need to **update the maps**\n",
    "There are also ~3-400 SNP discrepancy between 50k-v2, 777k and 50k-v3. So the following question is whether to use 50k-v3 or 777k map.\n",
    "Since [BovineHD](ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/productfiles/bovinehd/bovinehd-b1-annotation-file.zip) was created in 2014, [50k v3](https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/bovinesnp50/v3/bovinesnp50-v3-a2-manifest-file-csv.zip) was created in 2016, I will use 50k-v3 as the target map.\n",
    "\n",
    "- I am also considering to map these SNP with the most recent genome reference\n",
    "  - If I have time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update the 5 maps of the 6 platforms\n",
    "- [x] Extract the SNPs in column **3.** of **Table 1**\n",
    "  - Thre rest SNP in these 5 maps will discarded.\n",
    "  - The Norwegian data has no result from 50k-v3 platform. They can't be imputed to 51730 level of 50k-v3, as data from other countries can't be used as references. Let 777k SNP set as $v_7$, the Norwegian SNP set $s = \\cup(v_1, v_2, v_7)$, then the SNP in $\\cap(s, v_3)$ will be the final target number of SNP to impute to.\n",
    "  - Thus the shared number of loci will be even smaller. I summerized the final number of shared loci in **Table 2**\n",
    "- [x] Update the map with 50k-v3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Table 2**\n",
    "\n",
    "| Target | 50k-V1 | 50k-V2 |    HD | PF-10690 | PF-10993 | PF-11483 |\n",
    "| --: | --: | --: | --: | --: | --: | --: |\n",
    "|  51626 |  47877 |  49465 | 47155 |     7364 |     7438 |     7448 |\n",
    "\n",
    "Platform 50k v3 has 51730 autosomal SNP. Here I found 51626 usable. There is not a big loss at this stage."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new plink file set\n",
    "- New file sets are now in `data/plkmax`\n",
    "- The functions ran like a charm. SÃ¥ bra!\n",
    "- The Norwegian data seemed to have been manipulated already\n",
    "  - The final number is a little bit smaller than those in **Table 2**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge into 3 country sets\n",
    "- Found dozons of locuc pairs have the same chromosome numbers and bp positions\n",
    "  - [x] Create the pair list\n",
    "    - Total: 27 pairs\n",
    "    - are of same chromosome and base pair position\n",
    "  - [x] Check their manifest\n",
    "    - All the mutation flanking sequences are 60 bp long\n",
    "    - Visually checked: Manifests are basicly of same sequences, or reverse complements\n",
    "      - Some strange character like `YRKSWM`, rather than `ACGTN`\n",
    "  - [x] Check their genotypes\n",
    "    - Convert to VCF, and extract the corresponding genotype rows\n",
    "    - The genotypes are the same unless one is missing\n",
    "  - [x] Remove duplicates\n",
    "    - Keep names of only one column of the duplicated SNP pair\n",
    "    - Fill the genotypes if either one is not missing per ID.\n",
    "  - [x] Remove extra SNP from Dutch, and German data\n",
    "    - After above filter and merge, number of SNP left\n",
    "      | Dutch | German | Norge |\n",
    "      | --: | --: | --: |\n",
    "      | 51599 | 51599 | 51586|\n",
    "    - Shared across 3 countries: 51584\n",
    "  - [x] Remove non-shared across 3 countries.\n",
    "\n",
    "The result 3 sets will be subject to a final round of QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Further clean\n",
    "I do this in one step with a c++ program.\n",
    "- [x] Remove low MAF loci by country first\n",
    "  - Dutch $\\Rightarrow$ 45386\n",
    "  - German $\\Rightarrow$ 45656\n",
    "  - Norwegian $\\Rightarrow$ 46771\n",
    "- [x] Remove low quality ID. total loci: \n",
    "  - since the lowest number in **Table 2** is 7364\n",
    "  - I manually remove ID with less than 7000 loci of genotypes. Approx. 95%.\n",
    "  - Dutch 2602 $\\Rightarrow$ 2558\n",
    "    - Valid per ID $N_{\\mathrm{loci}}\\in[6139,\\,51495]$\n",
    "  - German 801 $\\Rightarrow$ 801.\n",
    "    - Minimum number of valid genotypes per ID, 48826. $\\frac{48826}{49465}=98.7$%\n",
    "    - Valid per ID $N_{\\mathrm{loci}}\\in[48826,\\,51545]$\n",
    "  - Norwegian 13491 $\\Rightarrow$ 7755. *Too many removed?*\n",
    "    - Refer **Table 2** $47877\\times .95\\approx 45483$\n",
    "    - Valid per ID $N_{\\mathrm{loci}}\\in[19322,\\,49412]$\n",
    "- [x] Remove non-shared loci across country \n",
    "  - 44015 loci left\n",
    "\n",
    "Below are to be determined.\n",
    "In summary, all on 44015 loci, missing or not, number of ID left:\n",
    "- Dutch: **2558**\n",
    "- German: **801**\n",
    "- Norwegian: **7755**\n",
    "\n",
    "I call this **dataset I**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove low quality loci\n",
    "This was forgot before the meeting on Apr. 29, 2020.\n",
    "Different thresholds of this filter should be different. Consider the platforms used in Dutch:\n",
    "\n",
    "| Platform | $N_{\\mathrm{ID}}$ | $N_{\\mathrm{loci}}$ |\n",
    "| --: | --: | --: |\n",
    "| 10690 | 1994 |  7364 |\n",
    "| 10993 |  102 |  7438 |\n",
    "| 11483 |  109 |  7448 |\n",
    "| 50kv2 |  343 | 49465 |                                                                                   | 50kv3 |   44 | 51626 |\n",
    "| 777k  |   10 | 47155 |\n",
    "\n",
    "Too many dutch platforms. Remove of low quality loci should be done before merging into 3 country sets.\n",
    "\n",
    "#### $N_{\\mathrm{loci}}$ left in German and Norwegian data\n",
    "Using the merged (of 3 pf) one:\n",
    "\n",
    "| Threshold | .1 | .2 | .3 | .4 | .5 | .6 | .7 | .8 | .9 |\n",
    "| -- | --: | --: | --: | --: | --: | --: | --: | --: | --: |\n",
    "| German | 51612 | 51609 | 51598 | 51598 | 51574 | 51567 | 51548 | 51529 | 51410 |\n",
    "| Norwegian | 51562 | 49468 | 49449 | 49415 | 48695 | 43691 | 43160 | 42708 | 38785 |\n",
    "\n",
    "Note `Threshold` is the minimal percentage of ID with valid genotypes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A plot above table:\n",
    "![](fig/rloci.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Summary of quality per locus\n",
    "- Hence the German data have the best quality about genotype missing per locus\n",
    "- Norwegian data are the worst.\n",
    "- I will remove low quality loci in Dutch and Norwegian data before merging into one set.\n",
    "  - using threshold 70%(?)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Beagle on above data.\n",
    "- [ ] Impute missing values in **dataset I**\n",
    "  - **Important**: $N_e$ for every country?\n",
    "  - use $N_e =100$"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
